### PR TITLE
Release/v2.7.0

### PR DESCRIPTION
**Pull request by twxia**
Include the following features:

- https://github.com/perpetual-protocol/perp-curie-contract/pull/87
- https://github.com/perpetual-protocol/perp-curie-contract/pull/86

### CODE REVIEW

**Comment by twxia** [comment]
Slither Summary
 - [uninitialized-state](#uninitialized-state) (1 results) (High)
 - [incorrect-equality](#incorrect-equality) (2 results) (Medium)
 - [reentrancy-no-eth](#reentrancy-no-eth) (3 results) (Medium)
 - [uninitialized-local](#uninitialized-local) (14 results) (Medium)
 - [unused-return](#unused-return) (2 results) (Medium)
## uninitialized-state
Impact: High
Confidence: High
 - [x] ID-0
[VaultStorageV1._totalDebt](contracts/storage/VaultStorage.sol#L22) is never initialized. It is used in:
	- [Vault.getTotalDebt()](contracts/Vault.sol#L306-L308)

contracts/storage/VaultStorage.sol#L22


## incorrect-equality
Impact: Medium
Confidence: High
 - [x] ID-1
[Vault._transferTokenIn(address,address,uint256)](contracts/Vault.sol#L601-L611) uses a dangerous strict equality:
	- [require(bool,string)((IERC20Metadata(token).balanceOf(address(this)).sub(balanceBefore)) == amount,V_IBA)](contracts/Vault.sol#L610)

contracts/Vault.sol#L601-L611


 - [x] ID-2
[MarketRegistry.addPool(address,uint24)](contracts/MarketRegistry.sol#L62-L101) uses a dangerous strict equality:
	- [require(bool,string)(IERC20Metadata(baseToken).balanceOf(_clearingHouse) == type()(uint256).max,MR_CHBNE)](contracts/MarketRegistry.sol#L68)

contracts/MarketRegistry.sol#L62-L101


## reentrancy-no-eth
Impact: Medium
Confidence: Medium
 - [x] ID-3
Reentrancy in [Exchange._swap(IExchange.SwapParams)](contracts/Exchange.sol#L443-L562):
	External calls:
	- [replayResponse = IOrderBook(_orderBook).replaySwap(IOrderBook.ReplaySwapParams(params.baseToken,marketInfo.pool,params.isBaseToQuote,true,signedScaledAmountForReplaySwap,params.sqrtPriceLimitX96,marketInfo.exchangeFeeRatio,marketInfo.uniswapFeeRatio,marketInfo.insuranceFundFeeRatio,fundingGrowthGlobal))](contracts/Exchange.sol#L458-L472)
	- [response = UniswapV3Broker.swap(UniswapV3Broker.SwapParams(marketInfo.pool,_clearingHouse,params.isBaseToQuote,params.isExactInput,scaledAmountForUniswapV3PoolSwap,params.sqrtPriceLimitX96,abi.encode(SwapCallbackData(params.trader,params.baseToken,marketInfo.pool,replayResponse.fee,marketInfo.uniswapFeeRatio))))](contracts/Exchange.sol#L476-L496)
	State variables written after the call(s):
	- [_firstTradedTimestampMap[params.baseToken] = _blockTimestamp()](contracts/Exchange.sol#L549)

contracts/Exchange.sol#L443-L562


 - [x] ID-4
Reentrancy in [Exchange.settleFunding(address,address)](contracts/Exchange.sol#L225-L279):
	External calls:
	- [IBaseToken(baseToken).cacheTwap(premiumInterval)](contracts/Exchange.sol#L239)
	- [fundingPayment = _updateFundingGrowth(trader,baseToken,IAccountBalance(_accountBalance).getBase(trader,baseToken),IAccountBalance(_accountBalance).getAccountInfo(trader,baseToken).lastTwPremiumGrowthGlobalX96,fundingGrowthGlobal)](contracts/Exchange.sol#L245-L251)
		- [liquidityCoefficientInFundingPayment = IOrderBook(_orderBook).updateFundingGrowthAndLiquidityCoefficientInFundingPayment(trader,baseToken,fundingGrowthGlobal)](contracts/Exchange.sol#L574-L579)
	State variables written after the call(s):
	- [(_lastSettledTimestampMap[baseToken],lastFundingGrowthGlobal.twPremiumX96,lastFundingGrowthGlobal.twPremiumDivBySqrtPriceX96) = (timestamp,fundingGrowthGlobal.twPremiumX96,fundingGrowthGlobal.twPremiumDivBySqrtPriceX96)](contracts/Exchange.sol#L261-L265)
	- [(_lastSettledTimestampMap[baseToken],lastFundingGrowthGlobal.twPremiumX96,lastFundingGrowthGlobal.twPremiumDivBySqrtPriceX96) = (timestamp,fundingGrowthGlobal.twPremiumX96,fundingGrowthGlobal.twPremiumDivBySqrtPriceX96)](contracts/Exchange.sol#L261-L265)

contracts/Exchange.sol#L225-L279


 - [x] ID-5
Reentrancy in [Vault._settleAndDecreaseBalance(address,address,uint256)](contracts/Vault.sol#L666-L688):
	External calls:
	- [IClearingHouse(_clearingHouse).settleAllFunding(to)](contracts/Vault.sol#L673)
	- [owedRealizedPnlX10_18 = IAccountBalance(_accountBalance).settleOwedRealizedPnl(to)](contracts/Vault.sol#L683)
	State variables written after the call(s):
	- [_modifyBalance(to,token,deltaBalance)](contracts/Vault.sol#L687)
		- [_balance[trader][token] = newBalance](contracts/Vault.sol#L723)
	- [_modifyBalance(to,token,deltaBalance)](contracts/Vault.sol#L687)
		- [collateralTokens[i] = collateralTokens[lastTokenIndex]](contracts/Vault.sol#L739)
		- [collateralTokens_scope_0.push(token)](contracts/Vault.sol#L747)

contracts/Vault.sol#L666-L688


## uninitialized-local
Impact: Medium
Confidence: Medium
 - [x] ID-6
[OrderBook.replaySwap(IOrderBook.ReplaySwapParams).insuranceFundFee](contracts/OrderBook.sol#L265) is a local variable never initialized

contracts/OrderBook.sol#L265


 - [x] ID-7
[AccountBalance.getTotalDebtValue(address).baseDebtValue](contracts/AccountBalance.sol#L233) is a local variable never initialized

contracts/AccountBalance.sol#L233


 - [ ] ID-8
[UniswapV3Broker.getNextInitializedTickWithinOneWord(address,int24,int24,bool).wordPos_scope_0](contracts/lib/UniswapV3Broker.sol#L280) is a local variable never initialized

contracts/lib/UniswapV3Broker.sol#L280


 - [x] ID-9
[OrderBook.replaySwap(IOrderBook.ReplaySwapParams).fee](contracts/OrderBook.sol#L264) is a local variable never initialized

contracts/OrderBook.sol#L264


 - [x] ID-10
[UniswapV3Broker.getNextInitializedTickWithinOneWord(address,int24,int24,bool).bitPos_scope_1](contracts/lib/UniswapV3Broker.sol#L280) is a local variable never initialized

contracts/lib/UniswapV3Broker.sol#L280


 - [x] ID-11
[OrderBook._getMakerQuoteBalanceAndPendingFee(address,address,bool).pendingFee_scope_0](contracts/OrderBook.sol#L675) is a local variable never initialized

contracts/OrderBook.sol#L675


 - [x] ID-12
[OrderBook._getTotalTokenAmountInPool(address,address,bool).amount](contracts/OrderBook.sol#L716) is a local variable never initialized

contracts/OrderBook.sol#L716


 - [x] ID-13
[Vault._modifyBalance(address,address,int256).i](contracts/Vault.sol#L735) is a local variable never initialized

contracts/Vault.sol#L735


 - [x] ID-14
[Exchange.swap(IExchange.SwapParams).pnlToBeRealized](contracts/Exchange.sol#L193) is a local variable never initialized

contracts/Exchange.sol#L193


 - [x] ID-15
[Tick.getAllFundingGrowth(mapping(int24 => Tick.GrowthInfo),int24,int24,int24,int256,int256).fundingGrowthRangeInfo](contracts/lib/Tick.sol#L86) is a local variable never initialized

contracts/lib/Tick.sol#L86


 - [x] ID-16
[ClearingHouse._modifyPositionAndRealizePnl(address,address,int256,int256,uint256,uint256).realizedPnl](contracts/ClearingHouse.sol#L703) is a local variable never initialized

contracts/ClearingHouse.sol#L703


 - [x] ID-17
[Exchange._getFundingGrowthGlobalAndTwaps(address).twapInterval](contracts/Exchange.sol#L629) is a local variable never initialized

contracts/Exchange.sol#L629


 - [x] ID-18
[Exchange.swap(IExchange.SwapParams).isPartialClose](contracts/Exchange.sol#L165) is a local variable never initialized

contracts/Exchange.sol#L165


 - [x] ID-19
[AccountBalance._deleteBaseToken(address,address).i](contracts/AccountBalance.sol#L462) is a local variable never initialized

contracts/AccountBalance.sol#L462


## unused-return
Impact: Medium
Confidence: Medium
 - [x] ID-20
[UniswapV3Broker.removeLiquidity(UniswapV3Broker.RemoveLiquidityParams)](contracts/lib/UniswapV3Broker.sol#L115-L133) ignores return value by [IUniswapV3Pool(params.pool).collect(params.recipient,params.lowerTick,params.upperTick,type()(uint128).max,type()(uint128).max)](contracts/lib/UniswapV3Broker.sol#L124-L130)

contracts/lib/UniswapV3Broker.sol#L115-L133


 - [x] ID-21
[InsuranceFund.repay()](contracts/InsuranceFund.sol#L69-L89) ignores return value by [IERC20Upgradeable(token).approve(vault,repaidAmount)](contracts/InsuranceFund.sol#L83)

contracts/InsuranceFund.sol#L69-L89

File: test/clearingHouse/ClearingHouse.customiseFee.test.ts [hunk 1]
```diff
@@ -736,6 +736,67 @@ describe("ClearingHouse customized fee", () => {
         })
     })
 
+    describe("change taker's fee discount ratio", async () => {
+        beforeEach(async () => {
+            // set fee ratio to 0.5%
```
Comment (42bchen) [inline test/clearingHouse/ClearingHouse.customiseFee.test.ts]: should be 2% ?

File: test/clearingHouse/ClearingHouse.openPosition.gasEstimation.ts [hunk 2]
```diff
@@ -84,6 +84,24 @@ describe.skip("ClearingHouse.openPosition gasEstimation", () => {
         })
     })
 
+    it("gas cost for taker", async () => {
```
Comment (42bchen) [inline test/clearingHouse/ClearingHouse.openPosition.gasEstimation.ts:87]: Should we remove `skip` if we want to add this test?

File: contracts/Exchange.sol [hunk 3]
```diff
@@ -53,6 +53,7 @@ contract Exchange is
     //
 
     struct InternalReplaySwapParams {
+        address trader;
```
Comment (42bchen) [inline contracts/Exchange.sol]: Although there is no security concern (this struct is only for internal functions), do we have any convention for adding new parameters to the structure?

File: contracts/interface/IOrderBook.sol [hunk 4]
```diff
@@ -41,12 +41,14 @@ interface IOrderBook {
...
         address baseToken;
+        address pool;
+        int256 amount;
         bool isBaseToQuote;
         bool shouldUpdateState;
-        int256 amount;
         uint160 sqrtPriceLimitX96;
         uint24 exchangeFeeRatio;
         uint24 uniswapFeeRatio;
+        uint24 insuranceFundFeeRatio;
```
Comment (42bchen) [inline contracts/interface/IOrderBook.sol]: same issue here

**Review by 42bchen — DISMISSED** [review]
LGTM, we can discuss some questions later

File: contracts/interface/IOrderBook.sol (hunk 4 above)
Comment (twxia) [reply contracts/interface/IOrderBook.sol]: f1ce6bb

File: test/clearingHouse/ClearingHouse.openPosition.gasEstimation.ts (hunk 2 above)
Comment (twxia) [reply test/clearingHouse/ClearingHouse.openPosition.gasEstimation.ts:87]: a07b42f

File: test/clearingHouse/ClearingHouse.customiseFee.test.ts (hunk 1 above)
Comment (twxia) [reply test/clearingHouse/ClearingHouse.customiseFee.test.ts]: a07b42f

File: contracts/Exchange.sol (hunk 3 above)
Comment (twxia) [reply contracts/Exchange.sol]: https://github.com/perpetual-protocol/perp-curie-contract/commit/f1ce6bb3f2b004d77257adc041eea2a308d94722

File: test/clearingHouse/ClearingHouse.openPosition.gasEstimation.ts (hunk 2 above)
Comment (twxia) [reply test/clearingHouse/ClearingHouse.openPosition.gasEstimation.ts:87]: I don't think we need it. This is just for checking purpose.

File: contracts/MarketRegistry.sol [hunk 5]
```diff
@@ -124,6 +130,11 @@ contract MarketRegistry is IMarketRegistry, ClearingHouseCallee, MarketRegistryS
         emit MarketMaxPriceSpreadRatioChanged(baseToken, ratio);
     }
 
+    function setFeeDiscountRatio(address trader, uint24 discountRatio) external checkRatio(discountRatio) onlyOwner {
```
Comment (paco0x) [inline contracts/MarketRegistry.sol:133]: can add a check that `discountRatio < _ONE_HUNDRED_PERCENT_RATIO`

Comment (42bchen) [reply contracts/MarketRegistry.sol:133]: We already have `checkRatio` modifier, but it didn't use `_ONE_HUNDRED_PERCENT_RATIO`, although these two value is the same

**Review by paco0x — DISMISSED** [review]

**Review by 42bchen — DISMISSED** [review]
GOGO

**Review by 42bchen — APPROVED** [review]

**Pull request merged**